build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -235,11 +257,46 @@ service ActionCache {
 // `INVALID_ARGUMENT` error will be returned. In either case, the client should
 // not attempt to retry the upload.
 //
-// For downloading blobs, the client must use the
-// [Read method][google.bytestream.ByteStream.Read] of the ByteStream API, with
-// a `resource_name` of `"{instance_name}/blobs/{hash}/{size}"`, where
-// `instance_name` is the instance name (see above), and `hash` and `size` are
-// the [Digest][build.bazel.remote.execution.v2.Digest] of the blob.
+// Small downloads can be grouped and requested in a batch via
+// [BatchReadBlobs][build.bazel.remote.execution.v2.ContentAddressableStorage.BatchReadBlobs].
+//
+// For large downloads, the client must use the
+// [Read method][google.bytestream.ByteStream.Read] of the ByteStream API.
+//
+// For uncompressed data, The `ReadRequest.resource_name` is of the following form:
+// `{instance_name}/blobs/{hash}/{size}`
+// Where `instance_name`, `hash` and `size` are defined as for uploads.
+//
+// Data can alternatively be downloaded in compressed form, with the following
+// `ReadRequest.resource_name` form:
+// `{instance_name}/compressed-blobs/{compressor}/{uncompressed_hash}/{uncompressed_size}`
+//
+// Where:
+// * `instance_name` and `compressor` are defined as for uploads.
+// * `uncompressed_hash` and `uncompressed_size` refer to the
+//   [Digest][build.bazel.remote.execution.v2.Digest] of the data being
+//   downloaded, once uncompressed. Clients MUST verify that these match
+//   the downloaded data once uncompressed, and take appropriate steps in
+//   the case of failure such as retrying a limited number of times or
+//   surfacing an error to the user.
+//
+// Note that when reading compressed blobs, the `ReadRequest.read_offset`
+// refers to the offset in the uncompressed form of the blob, and the
+// `ReadRequest.read_limit` refers to the offset in the specific compressed
```

Comment:
I don't quite understand this comment. read_limit is not an offset akin to read_offset, it's a limit on the number of bytes to be read _from_ that point. (I read this sentence as implying offset from beginning, though I'm not sure if that's how it was intended). 

When would a caller want to set a read_limit to constrain the number of compressed bytes returned? I suppose I could imagine it as a defensive mechanism to avoid unbounded reading (since technically the compressed size could be higher than the uncompressed size), but the client can also just abort the read if they're getting too many raw bytes back. But it seems like it'd be more semantically meaningful for it to also refer to the uncompressed bytes, so that one can express "read me 1MB of bytes starting at offset 100MB", independent of compression.

Reply:
> I don't quite understand this comment. read_limit is not an offset akin to read_offset, it's a limit on the number of bytes to be read _from_ that point. (I read this sentence as implying offset from beginning, though I'm not sure if that's how it was intended).

They are only akin in the sense that they're siblings in the ReadRequest message which I think deserve clarification for our use case. Perhaps we should split these into distinct sentences, or dot points?

While it would be easier to understand from the client side if read_limit was specified in uncompressed bytes, I don't think it's possible to implement well on the server side without sacrificing more important qualities or only providing an approximate limit (eg at the chunk-size resolution if the server implements that).

> When would a caller want to set a read_limit to constrain the number of compressed bytes returned? I suppose I could imagine it as a defensive mechanism to avoid unbounded reading (since technically the compressed size could be higher than the uncompressed size), but the client can also just abort the read if they're getting too many raw bytes back. But it seems like it'd be more semantically meaningful for it to also refer to the uncompressed bytes, so that one can express "read me 1MB of bytes starting at offset 100MB", independent of compression.

I suspect read_limit is something that is useful for other bytestream use cases, eg a defensive mechanism as you mentioned, when the size of the thing being requested isn't specified in the resource_name. I do not think it's useful for us here.

I would be OK with instead stating that server should ignore this field. Should we do that?

Reply:
Oh wait- I meant for this new text to say something like "refers to a limit of the number of compressed bytes in the specific compressed form."

Reply:
> They are only akin in the sense that they're siblings in the ReadRequest message which I think deserve clarification for our use case. Perhaps we should split these into distinct sentences, or dot points?

Dot points like the other arguments above sound useful.

I see a couple other semantic problems with having `read_limit` refer to the uncompressed size:
* In that case, Bytestream.Read now has to be aware of two limits: the one to advertise in the request, and the one to enforce for aborting the connection.
* If the client is aborting the connection as a way of limitting, `read_limit` becomes a somewhat redundant argument when reading _uncompressed_ bytes.

Reply:
> While it would be easier to understand from the client side if read_limit was specified in uncompressed bytes, I don't think it's possible to implement well on the server side without sacrificing more important qualities or only providing an approximate limit (eg at the chunk-size resolution if the server implements that).

It would be a bit fiddly, but not atrocious: uncompress, truncate, and recompress the final block, similar to what we have to do for the initial block of an offset read. I can definitely see the argument it's not worth the effort though.

Rubens:
> In that case, Bytestream.Read now has to be aware of two limits: the one to advertise in the request, and the one to enforce for aborting the connection.

I'm not sure I understand this point - are you referring to a client implementation of Read? Presumably a client will pick one or the other and not both.

> If the client is aborting the connection as a way of limitting, read_limit becomes a somewhat redundant argument when reading uncompressed bytes.

TBH, whether you want to limit on compressed bytes or uncompressed bytes, you can do it by just reading till you've gotten enough and then aborting. Which is maybe an argument towards Mostyn's proposal of just ignoring the field? 

The only use-case I actually have in mind for read_limit is UIs - where e.g. "Read 1MB at offset 100MB" is theoretically useful for paginating large files, and would require the limit to be expressed in terms of uncompressed bytes to be useful. Limiting compressed bytes is kinda bizarre - it either requires you to guess how many compressed bytes will be needed to get the right number of uncompressed bytes, or to arbitrarily truncate the read after some arbitrary number of logical bytes have been read, as a way to defend against...something. (But it'll decompress into arbitrarily large output, so not a very good defence against memory blowup?).

But again, I don't know that it matters enough - if you have a preference here, SGTM.

Reply:
If we don't want servers to implement this, should we require servers to return INVALID_ARGUMENT if the value is non-zero, or to ignore the field completely? I think the former would be easier to gracefully migrate from if we decide to add support later.

Reply:
> I'm not sure I understand this point - are you referring to a client implementation of Read? Presumably a client will pick one or the other and not both.

I do mean a client implementation of Read. If `ByteStream.ReadRequest.read_limit` refers to the uncompressed bytes, and the Read call will abort after a certain number of *raw* bytes, that's two limits to be aware about. I guess you could argue that this is implementation specific, or not exactly an issue, but this would likely means refactoring of some existing bytestream client implementations to support this - the [go remote-apis-sdks](https://github.com/bazelbuild/remote-apis-sdks/blob/master/go/pkg/client/bytestream.go#L101) being one of them. I'm not sure if that's a significant issue.

I think that, plus having to rework the last block of a blob in the server side, make it slightly harder to support `read_limit` referring to the uncompressed bytes. But I also agree that this control can be much more useful.

> If we don't want servers to implement this, should we require servers to return INVALID_ARGUMENT if the value is non-zero, or to ignore the field completely? I think the former would be easier to gracefully migrate from if we decide to add support later.

Being explicit with INVALID_ARGUMENT sounds better than silent ignores IMO.

Reply:
Pushed a fixup that reorganises the "when reading compressed-blobs" statements into dot points, and require servers to return INVALID_ARGUMENT for non-zero read_limit values.